tests/cypress/integration/basic.js

```diff
@@ -0,0 +1,7 @@
+// The simplest tests, validating that it's possible to render the app at all.
+describe('landing page', () => {
```

Comment:
Jak będziemy provisionować stan aplikacji (faza setup), skoro JavaScript?

Reply:
Jeśli mam być szczery, to moje doświadczenie z cypress nie jest duże, ale wiem że istnieje możliwość wywołania dowolnej komendy z użyciem [cy.exec](https://docs.cypress.io/api/commands/exec.html#Syntax). W takiej sytuacji kod testu (najprawdopodobniej) będzie mógł komunikować się bezpośrednio z bazą poprzez dedykowane skrypty.

Reply:
Ja także nie mam dużego doświadczenia w testowaniu front-endu, a w zakresie aplikacji Python, a testy Javascript nie mam doświadczenia żadnego. Jeżeli będziemy używać `cy.exec` to pozostanie kwestia odpowiedniego zarządzania tymi skryptami (napisanymi w Pythonie). Dołożyłeś nowy kontener bazujący na `cypress`, więc wywołanie z niego Pythona nie będzie trywialne.

Reply:
Raczej miałem na myśli wywołanie skryptów napisanych też w JS, gadajacych bezpośrednio z bazą danych. Testy byłyby zdefiniowane w tym samym kontenerze co testy.

Generalnie nie spodziewam się, przynajmniej póki co, wywoływania niczego bardziej skomplikowanego niż usunięcie wszystkich danych przed testem, ewentualnie stworzenie możliwości dodania paru kluczowych obiektów. Raczej parę podstawowych, łatwych w utrzymaniu skryptów JS.

Reply:
Możemy spróbować, ale rezygnowanie z fabryk, które mamy na back-endzie wywołuje moje obawy w zakresie kosztów rozwoju i utrzymania. Obecnie większość stanu jesteśmy w stanie zrobić jednym wywołaniem fabryki, ewentualnie odpowiednim wywołaniem ORM-a. Uważam to za dużą wartość i ułatwienie w pisaniu testów. Tu proponujesz zejście z tych warstw na najniższy poziom.

Co zdecydowało o wyborze Cypress do testów? Rozważałeś klasyczne Selenium, które mniej modne, ale mogłoby nam zapewnić bliższą integracje z tym co dotychczas mamy ( https://blog.mloza.pl/testy-w-django-z-uzyciem-selenium/ ), albo coś w Pythonie np. Pyppeteer / Splinter?

Reply:
Cóż, główny problem z Selenium jest taki, że to bardzo stara technologia, więc o ile byłoby to prostsze w kwestii pisania samych testów (Python), to spodziewałbym się problemów z utrzymaniem, z racji wieku biblioteki.

Wydaje mi się, że problemem tutaj nie jest kwestia użycia Cypress / Selenium, ale kwestia umieszczenia testów w oddzielnym kontenerze. Z cypress mamy dostęp do basha, więc gdyby był w tym samym kontenerze co django, mógłby po prostu wywołać dowolny skrypt py i rozwiązać problem.
Wydaje mi się jednak, że umieszczenie testów integracyjnych w oddzielnym kontenerze ma więcej zalet niż wad. Koniec końców, chcemy by testy były jak najbliższe rzeczywistości, więc komunikacja bezpośrednio z django nie powinna być opcją.
Co do samych testów - raczej zakładam, że doprowadzenie aplikacji do konkretnego stanu powinno być częścią samego testu, a nie kodu wykonywanego przed nim, tj.:
- jeśli chcemy testować edycję spraw, w ramach testu powinniśmy wywołać
  1. logowanie / rejestrację użytkownika
  2. utworzenie kilku spraw
  3. aktualizację konkretnej sprawy

To oczywiście zwiększy czas trwania testu, ale wykryje więcej potencjalnych błędów. Nie mamy pewności, czy wywołując ręcznie factory django nie pomijamy czegoś co wydarza się po przeklikaniu w przeglądarce, a to właśnie chcemy przetestować.

Ręczna komunikacja z bazą danych (i ogólnie z bashem) mogłaby ograniczyć się do:
1. sprzątania bazy po każdym teście
2. ewentualnego ustawienia wartości, których obecnie nie można przeklikać (np. teryt)
3. obsługi komunikacji, która w praktyce odbywa się poza przeglądarką (np. email - nie wiem dokładnie jak wygląda implementacja, ale zakładam, że - przynajmniej w jakimś stopniu - można zastąpić emaile bezpośrednią komunikacją z bazą).
Innymi słowy, rzeczami, które nie będą (za bardzo) wymagały znajomości modeli Django.